### ISSUE TITLE
establish several connection string in one macro

### ISSUE THREAD

**Issue by chifaanfar** [Seddryck/NBi#362]
Hello 
I'm trying to test the existing of a dimension in two differents cubes , so i have to establish two connection string , i think that reference can be helpful , so i try to use the setting reference , but i don't know how to work with it , in a genbiL file :
```
setting reference 'referenceName' connectionString 'data source=localhost; initial catalog=myDb';
```
my question is : is it possible in genbiL , to establish more than one connection string , if yes ,  how to do it? .

thank you for your return in advance .

**Comment by Seddryck** [comment]
Using references is the way to go when you've more than one connection-string for your system-under-test. In genbil, to create two references, you'll need to use twice the command ```setting reference```.

```
setting reference 'reference1' connectionString 'data source=localhost; initial catalog=myDb;Trusted Connection = true;'
setting reference 'reference2' connectionString 'data source=myOtherServer; initial catalog=myOtherDb;Trusted Connection = true;'
```
Then in your tests, you'll need to reference one or the other reference:
```xml
<query connectionString="@reference1">...</query>
```

On the other hand, if you want to apply the same set of tests to two distinct cubes, you haven't to duplicate your tests, you've two create two config files with different connection-strings. In fact, you can specify in your test to use a reference or a default from the settings ... and in the settings you can specify that the connection-string is in fact in the config file! More info at www.nbi.io/docs/connection-configuration/

**Comment by chifaanfar** [comment]
thanks for your return
you mean that after writing the references in my genbiL file , i have to specify for each test the "@reference", but the question is , how can i automate this specification , because tests are generated automaticaly with GenbiL 
as you can see in the pictures , i specified the two connection string in the macro , but i can't automates the name of reference to give for each test .
I'm stuck in this part for days :(
i hope that i was clear 

![test](https://user-images.githubusercontent.com/37936281/38491808-82c700a4-3bdc-11e8-8997-c8fffc09f815.PNG)
![test2](https://user-images.githubusercontent.com/37936281/38491813-865a06bc-3bdc-11e8-919a-0f5ab996740e.PNG)

**Comment by Seddryck** [comment]
Based on your comment, I'd really check if what you're trying to achieve is not easier by having two config files pointing to the same test-suite.

But if you want to continue to deal with two connection-strings in references and one test-suite then you'll need to play with your ```case``` and a [cross](http://www.nbi.io/automation/cases-cross/).

To have each test created for the first reference and also for the second reference, you'll have to execute:
```
case scope 'temps';
case load file '...\temps.csv';
case cross 'temps' with vector 'ref' values 'ref-one', 'ref-two'
```
Basically, if you've 5 entries (lines) in your file 'temps.csv' then your case will now contains 10 entries (5 associated to ref-one and five to ref-two).

Then you'll need to create a template where the connection string can be specified

```xml
...
<dimension ... connectionString="@$ref$"/>
...
```
And load this template with
```
template load file '..\dimensionExists.nbitt';
```

**Comment by chifaanfar** [comment]
thank you seddryck , it works better now , but with two csv files , each csv file connected to a reference.

i have another point to ask you for (I have little time to finish this project and pass my defense, so I try to ask the maximum of questions to move forward).

the most test that i have to automatised  consists in a validation that dimension or hierarchy or mesures are displayed in a specified order , but according to the documentation , it's just available with members  http://www.nbi.io/docs/members-ordering/ 
there is a solution to propose to me , you will make me a great pleasure .

**Comment by Seddryck** [comment]
Not sure that I understand the question but the  ordering of dimensions or measures is usually client-specific. You can’t specify this on the server-side (reason why most of the clients are just displaying them alphabetically). NBi can’t test the clients.

**Comment by chifaanfar** [comment]
for exemple , for dimension('temps') , we want to test if the hierarchy is on a numeric order , or alphabetic , and sometimes we want to display the dimension/measure in a specific order , in general we want to apply the possibility of classification offered to the members for the dimensions.
![test3](https://user-images.githubusercontent.com/37936281/38510409-86f509c2-3c14-11e8-8c68-ad3b33421c3b.PNG)

**Comment by Seddryck** [comment]
Then in that case you need to use [members’ ordering](http://www.nbi.io/docs/members-ordering/). You’ll be able to check that the members are ordered “lundi” , “mardi”, “mercredi” ... et “janvier”, “février”, “mars” ...

**Comment by chifaanfar** [comment]
yeah i see it before ,this section about members' odrering , and i would like to apply this options on the dimensions and their hierarchy .
so , Based on your answers , i can see that i don't have this possibility of ordering , it's available just for members :(

**Comment by Seddryck** [comment]
I don’t think that this is not testable with NBi. Based on your previous screenshot could you give me a concrete example of what you want to test? (If needed expand your hierarchies).

**Comment by chifaanfar** [comment]
by exemple we have received a test request to automate and the aim is to have the dimension displayed in alphabetical order, as you can see, they are not ranked well, [Temps avant Nature d'appel] 

 and sometimes we want to display the dimension in a specific order [ for exemple having the dimension Temps displayed at the beginning]
![test3](https://user-images.githubusercontent.com/37936281/38569767-b46c8052-3cdb-11e8-9eda-bb4c825cfd58.PNG)

**Comment by Seddryck** [comment]
You cannot put a constraint to sort the dimensions with SSAS. It means that you cannot guarantee that the dimension "Géographie" will or will not be listed before the dimension "Service" in SSAS. Just ask the dev which settings she is using to specify that dimension "Nature d'appel" is displayed after dimension "Temps" ... the answer would probably be "none ... it's just random". Sorting of dimensions is not managed on the server-side but on the client-side. It means that you shouldn't test the cube (server-side) but the tool that you'll use to connect to it (client-side). NBi is focussed on testing the server-side. **Globally the requirement that you receive is not implementable in SSAS and must be forwarded to the client.**

Most of the clients (I can't find exceptions) are displaying dimensions in alphabetical order. Don't forget that you can also translate dimension's names ... that could complexify how clients are ordering the dimensions. 

If you built your own client (or you're dealing with a weird client) then it could use the DIMENSION_ORDINAL attribute. This attribute cannot be set in Visual Studio so it would be really weird but who knows, I've seen some really weird ideas. In that specific case, the best test that you can automate with NBi is something such as:

```xml
<system-under-test>
  <resultSet>
    <query connectionString="@ssasMD">
       SELECT DIMENSION_ORDINAL, DIMENSION_CAPTION
       FROM $System.MDSCHEMA_DIMENSIONS
       WHERE CUBE_NAME='Adventure Works'
       ORDER BY DIMENSION_ORDINAL
    </query>
  </resultSet>
</system-under-test>
<assert>
  <equalTo>
     <resultSet>
        <row><cell>1</cell><cell>Measure</cell></row>
        <row><cell>2</cell><cell>Date</cell></row>
        <row><cell>3</cell><cell>Ship Date</cell></row>
     </resultSet>
  </equalTo>
</assert>
```

**Comment by chifaanfar** [comment]
thanks you for this clarification , i can understand now , thank you again
